### ISSUE TITLE
kvscheduler internal graph bug

### ISSUE THREAD

**Issue by scottlx** [ligato/vpp-agent#1865]
I found a bug when I implemented my plugin. The descriptor is quite simple, which only depends on another descriptor. However, deleting the kv occasionally cause the following panic:

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xc9b604]

goroutine 23 [running]:
go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.(*node).removeFromSources(0xc0008a2570, {0x1b33afb, 0xc0007d64e0}, {0x1b35dcb, 0xc}, {0xc0007d6500, 0x1d})
	/src/ligato/vpp-agent/plugins/kvscheduler/internal/graph/node_write.go:373 +0x84
go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.(*node).SetTargets(0xc0008a25f0, {0xc0007b64b0, 0x1d, 0x0})
	/src/ligato/vpp-agent/plugins/kvscheduler/internal/graph/node_write.go:188 +0xb8f
go.ligato.io/vpp-agent/v3/plugins/kvscheduler.(*Scheduler).applyNewRelations(0xc000025a08, {0x1e38fb0, 0xc0008a25f0}, 0x1da3c80, {0x0, 0x0}, 0x1, 0xc00085e0b0)
	/src/ligato/vpp-agent/plugins/kvscheduler/txn_exec.go:705 +0x2e2
go.ligato.io/vpp-agent/v3/plugins/kvscheduler.(*Scheduler).applyCreate(0x3051b80, {0x1e38fb0, 0xc0008a25f0}, 0xc0009ffb80, 0xc00085e0b0)
	/src/ligato/vpp-agent/plugins/kvscheduler/txn_exec.go:427 +0x5e7
go.ligato.io/vpp-agent/v3/plugins/kvscheduler.(*Scheduler).applyValue(0x3051b80, 0xc00085e0b0)
	/src/ligato/vpp-agent/plugins/kvscheduler/txn_exec.go:241 +0x81c
go.ligato.io/vpp-agent/v3/plugins/kvscheduler.(*Scheduler).executeTransaction(0x3051b80, 0xc001440060, {0x1e38c70, 0xc0017971d0}, 0x0)
	/src/ligato/vpp-agent/plugins/kvscheduler/txn_exec.go:85 +0xa65
go.ligato.io/vpp-agent/v3/plugins/kvscheduler.(*Scheduler).processTransaction(0x3051b80, 0xc001440060)
	/src/ligato/vpp-agent/plugins/kvscheduler/txn_process.go:152 +0x2bc
go.ligato.io/vpp-agent/v3/plugins/kvscheduler.(*Scheduler).consumeTransactions(0x3051b80)
	/src/ligato/vpp-agent/plugins/kvscheduler/txn_process.go:97 +0xe5
created by go.ligato.io/vpp-agent/v3/plugins/kvscheduler.(*Scheduler).Init
/src/ligato/vpp-agent/plugins/kvscheduler/plugin_scheduler.go:197 +0x450

```

It seems like sometimes the sources field of node is set to nil in other places.

```shell
	p node.graph.nodes["config/saasfw/group/v1/cfg/45"]
	*go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.node {
		nodeR: *go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.nodeR {
			graph: *(*"go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.graphR")(0xc0005982c0),
			key: "config/saasfw/group/v1/cfg/45",
			label: "45",
			value: google.golang.org/protobuf/reflect/protoreflect.ProtoMessage(*go.ligato.io/vpp-agent/v3/proto/ligato/saasfw/group.Group) ...,
			flags: [8]go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.Flag [
				...,
				nil,
				...,
				...,
				...,
				nil,
				nil,
				nil,
			],
			metadata: interface {} nil,
			metadataAdded: false,
			metadataMap: "saasfw-group",
			targets: go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.Targets len: 1, cap: 1, [
				(*"go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.Target")(0xc0013503c0),
			],
			targetsDef: []go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.RelationTargetDef len: 1, cap: 1, [
				(*"go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.RelationTargetDef")(0xc0018aa1e0),
			],
			sources: go.ligato.io/vpp-agent/v3/plugins/kvscheduler/internal/graph.Targets len: 0, cap: 1, [],},
		metaInSync: true,
		dataUpdated: true,
		targetsUpdated: false,
		sourcesUpdated: false,}
```

Should skip nil value here:

https://github.com/ligato/vpp-agent/blob/74793c7e4a334fb2364215995d5e4020456e89e3/plugins/kvscheduler/internal/graph/node_write.go#L372

**Comment by ondrej-fabry** [comment]
Hey, thanks for the issue report. Could you share your descriptor implementation? I would like to understand why this is happening. 

By the way, if you managed to implement a fix, you are welcome to open a PR.

**Comment by scottlx** [comment]
The descriptor depends on another two resource. All descriptors don't have derives and retrieve method. Instead of calling govpp, create method sends abf and acl message to local client and triggers another kv transcation.

```go
func NewPropertyDescriptor(
	log logging.Logger,
	abf abfvppcalls.ABFVppAPI,
	acl aclvppcalls.ACLVppAPI,
	vrIndex vrouteridx.VrouterMetadataIndex,
	grIndex groupidx.GroupMetadataIndex,
	converter *groupconverter.GroupConverter,
	externalInterface string,
) *api.KVDescriptor {
	ctx := &PropertyDescriptor{
		log:               log,
		abfHandler:        abf,
		aclHandler:        acl,
		vrouterIndexes:    vrIndex,
		groupIndexes:      grIndex,
		externalInterface: externalInterface,
		converter:         converter,
	}
	typedDescr := &adapter.PropertyDescriptor{
		Name:               PropertyDescriptorName,
		KeySelector:        property.ModelProperty.IsKeyValid,
		ValueTypeName:      property.ModelProperty.ProtoName(),
		KeyLabel:           property.ModelProperty.StripKeyPrefix,
		ValueComparator:    ctx.EquivalentProperty,
		NBKeyPrefix:        property.ModelProperty.KeyPrefix(),
		WithMetadata:       true,
		MetadataMapFactory: ctx.MetadataFactory,
		Validate:           ctx.Validate,
		Create:             ctx.Create,
		Delete:             ctx.Delete,
		Update:             ctx.Update,
		Dependencies:       ctx.Dependencies,
	}
	return adapter.NewPropertyDescriptor(typedDescr)
}

func (d *PropertyDescriptor) Dependencies(key string, propertyData *property.Property) (dependencies []api.Dependency) {
	// group
	groupDepends := make(map[uint32]api.Dependency)
	for _, target := range propertyData.ProtectTargets {
		for _, chain := range target.AttachedGroup {
			groupDepends[chain.GroupId] = api.Dependency{
				Label: "group-exists",
				Key:   group.GroupKey(chain.GroupId),
			}
		}
	}
	for _, dependency := range groupDepends {
		dependencies = append(dependencies, dependency)
	}

	//vrouter
	dependencies = append(dependencies, api.Dependency{
		Label: "vrouter-exists",
		Key:   vrouter.VrouterKey(vrouterdescriptor.VrouterUniqueId),
	})
	return dependencies
}

func (d *PropertyDescriptor) Create(key string, propertyData *property.Property) (*idxvpp.OnlyIndex, error) {
	metadata := &idxvpp.OnlyIndex{
		Index: PropertyUniqueIndex,
	}
	msg, err := d.generateMessage(propertyData, nil)
	if err != nil {
		return nil, err
	}
	if err = client.LocalClient.ChangeRequest().Update(msg...).Send(context.Background()); err != nil {
		return nil, errors.Errorf("failed to create vrouter config: %v", err)
	}

	return metadata, nil
}
```

**Comment by ondrej-fabry** [comment]
The KVScheduler processes all transactions synchronously. Triggering the KVScheduler from the methods of KVDescriptor (Create, Delete..) should cause a dead lock because the call would not be able to finish - it would be practically waiting for itself.

This is understandably a good use case, but should not be possible because of the KVScheduler limitation I mentioned above. I am not even sure how that works for you.. can you share log?

**Comment by scottlx** [comment]
Here is how the code works:

1. In transaction1, kvscheduler runs my descriptor's create method.
2. The create method calls LocalClient.ChangeRequest().Update().Commit(), which create transcation2 and enque.
3. The scheduler blocks at here:

https://github.com/ligato/cn-infra/blob/012faf45555e52dc5bf71a6adc9680b5bcd30665/datasync/syncbase/watcher.go#L156

5. Because the scheduler is still processing transaction1, it won't start transaction2. After PropagateChangesTimeout, transaction1 will finish.

My environment is in company intranet, but the log looks like this:
```shell
==================
transaction 1  
create my descriptor  xxxxx
==================
Push data with 3 KV pairs (source: %s)", len(kvPairs), dataSrc
- PUT acl
- PUT abf
- PUT route
Timeout of aggregated data-change callbacks 20s
Transaction #1 successful! (took xx)

==================
transaction 2 
create acl, abf, route xxxxx
==================
...
Transaction #2 successful! (took xx)
```

It is definitely not the best way to write a plugin. At the very first, I tried using derive method. But I found if the derived value has another derived value, the last derived value won't send to vpp. For example, my descriptor derives interface, and interface derives interface_address, interface_with_address etc.  Interface_address and interface_with_address does not take effect.

Today I did lots of testing, but the panic did not occur again. Quite weird... It seems like the node structure is not thread safe. I suggest adding nil judgement for the target in removeFromSources method because node.Source could be empty.
https://github.com/ligato/vpp-agent/blob/d830538a5bbad58cef3ab980c20d866597539bee/plugins/kvscheduler/internal/graph/node_write.go#L372

Maybe it is cause by a loop? or time racing? Anyway, I'm going to close this issue for it is not a typical circumstance.

Thanks for your awesome code. It really helps me a lot! ^-^
